documentation/preview-push.md

```diff
@@ -96,18 +97,31 @@ OPTIONS:
     performed corrections in the file named `name`. If no name is specified, no
     report is generated.
 
-## ppreview/ppush
-
-{% hint style="info" %}
-Starting in v4.9
-{% endhint %}
-
-The `ppreview`/`ppush` subcommands are a preview of a future feature where zone
-data is gathered concurrently. The commands will go away when
-they replace the existing `preview`/`push` commands.
-
 * `--cmode value`
   * Concurrency mode. Specifies what kind of providers should be run concurrently.
     * `default` -- Providers are run sequentially or concurrently depending on whether the provider is marked as having been tested to run concurrently.
-    * `none` -- All providers are run sequentially. This is the safest mode. It can be used if a concurrency bug is discovered.
-    * `all` -- This is unsafe. It runs all providers concurrently, even the ones that have not be validated to run concurrently. It is generally only used for demonstrating bugs.
+    * `none` -- All providers are run sequentially. This is the safest mode. This is useful if concurrency bugs are discovered.
+    * `all` -- This is unsafe. It runs all providers concurrently, including the ones that have not be validated to run concurrently. This is useful for demonstrating bugs.
+
+## Concurrency
+
+The first phase of `preview`/`push` is to gather information about the zones.
+This can be quite time consuming. Each zone's information is gathered
+concurrently, when possible, to reduce run-time. However not all providers have
+been throughly tested to verify they support concurrent operation. Zones from
+those providers are run sequentially after the concurrent providers complete.
+The `--cmode` flag can be used to force zones to run sequentially or
+concurrently.
+
+Originally zone information was gathered sequentially.  In v4.9.0 concurrent
+gathering was introduced via new commands: `ppreview`/`ppush`.  In v4.14.0
+concurrency became the default: `preview`/`push` became aliases for `ppreview`/`ppush`
+the original sequentially commands were renamed `oldpreview`/`oldpush`.
+In the future the `oldpreview`/`oldpush` code will disappear.
+
+| Version    | `preview`/`push` | `oldpreview`/`oldpush` | `ppreview`/`ppush` |
+|------------|------------------|------------------------|--------------------|
+| pre-v4.9   | sequential code  |  n/a                   | n/a                |
+| v4.9       | sequential code  |  n/a                   | concurrent code    |
+| v4.14      | concurrent code  | serial code            | concurrent code    |
+| future     | concurrent code  | removed                | concurrent code    |
```

Comment:
I wonder, why not, in accordance with Semantic Versioning, introduce a v5 where the old `preview`/`push` commands are removed? 🤔 This saves us the hassle of drawing up overviews. Of course we will include this in the next release notes as a noticeably ('breaking'?) change.

Reply:
You mean change "future" to "v5"?

Reply:
The (Markdown) tables as included in this GitHub pull request.

Reply:
Ah, I see.

Here's what I'm trying to achieve:
1. I want to encourage people to use the new code; and discourage people from using the old code.
2. I want to remove the old code as soon as possible (maybe 2-3 months max).

What do you recommend?

Reply:
I would recommend:

1. **Release v4.14** (<1 week)

   * With a deprecation message within the `preview` and `push` commands.
      * Context: 'in de next release (`v4.15`) the parallel function (`ppreview` and `ppush`) will be the default and so enabled for all providers. We encourage you to test the new `ppreview` and `ppush` commands for your own DNSControl set-up/configured providers.'
   * Add a (similar) deprecation message to the v4.14 release notes.
   * Send a mailing to the mailinglist with the release notes (deprecation message included).

2. **Release v4.15** (1 month)

    * In release `v4.15` change the default behavior of the  `preview` and `push` commands to the 'parallel version' (`ppreview` and `ppush`) and while keeping the older serially version of `preview` and `push` functionalities as fallback. Just like GitHub pull request https://github.com/StackExchange/dnscontrol/pull/3132 does.

3. **Release v5.0** (~3 months)

    * In release `v5.00` remove the serially `preview` and `push` commands.
    * _Another nice removal for v5_: https://github.com/StackExchange/dnscontrol/blob/d6d50fc8625f893454872668a37bf26cdbedbd3c/commands/ppreviewPush.go#L178

I hope this helps you further in creating a release plan. If you have any questions, please let me know.

Reply:
Excellent!  Yes, that sounds good.

Should `oldpreview`/`oldpush` appear earlier i.e. in v4.14?

Reply:
Sequence looks good but I would think you can pull in the timelines if you want since v5 signals a major change and folks can decide when they want to roll it out internally,

Reply:
> Should `oldpreview`/`oldpush` appear earlier i.e. in `v4.14`?

No, I would roll back this change https://github.com/StackExchange/dnscontrol/pull/3132 in the `main` branch and (re)apply it in a later release `v4.15`. This gives the users another chance to prevent ignorance. After all, there will be (to be realised) a deprecation/log entry in the old `preview` and `push` commands.

> Sequence looks good but I would think you can pull in the timelines if you want since v5 signals a major change and folks can decide when they want to roll it out internally,

That would certainly be possible. I would let the timeline depend mainly on the available time. @tlimoncelli himself spoke about 2/3 months where I assumed to prepare the various activities. In theory you could release `v4.15` and `v5.00` at the same time because it applies Semantic Versioning.

Reply:
1. I propose we add oldpreview/oldpush as aliases for preview/push in v4.14. It dosn't hurt to make these commands available early.

2. The docs for oldpreview/oldpush should warn that these aliases are not subject to SemVer and should only be used with the understanding that they'll go away.

3. I disagree that we need to remove the old code in a major release (v5).  Feature-wise, there's no breaking change between oldpreview and ppreview.   The only breakage would be if someone uses oldpreview/oldpush in a script; my solution for that is in the previous paragraph.

Reply:
Let's move this to here: https://github.com/StackExchange/dnscontrol/issues/3142